cpp/src/neighbors/ann_mg/ann_mg.cuh

```diff
@@ -0,0 +1,747 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../detail/knn_merge_parts.cuh"
+#include <raft/core/serialize.hpp>
+#include <raft/util/cuda_dev_essentials.cuh>
+
+#include "nccl_helpers.cuh"
+#define NO_NCCL_FORWARD_DECLARATION
+#include <cuvs/neighbors/ann_mg.hpp>
+#undef NO_NCCL_FORWARD_DECLARATION
+#include <cuvs/neighbors/common.hpp>
+
+namespace cuvs::neighbors::mg {
+using namespace cuvs::neighbors;
+using namespace raft;
+
+template <typename AnnIndexType, typename T, typename IdxT>
+template <typename Accessor>
+void ann_interface<AnnIndexType, T, IdxT>::build(
+  raft::resources const& handle,
+  const cuvs::neighbors::index_params* index_params,
+  raft::mdspan<const T, matrix_extent<IdxT>, row_major, Accessor> index_dataset)
+{
+  if constexpr (std::is_same<AnnIndexType, ivf_flat::index<T, IdxT>>::value) {
+    auto idx = cuvs::neighbors::ivf_flat::build(
+      handle, *static_cast<const ivf_flat::index_params*>(index_params), index_dataset);
+    index_.emplace(std::move(idx));
+  } else if constexpr (std::is_same<AnnIndexType, ivf_pq::index<IdxT>>::value) {
+    auto idx = cuvs::neighbors::ivf_pq::build(
+      handle, *static_cast<const ivf_pq::index_params*>(index_params), index_dataset);
+    index_.emplace(std::move(idx));
+  } else if constexpr (std::is_same<AnnIndexType, cagra::index<T, IdxT>>::value) {
+    auto extents = raft::make_extents<int64_t>(index_dataset.extent(0), index_dataset.extent(1));
+    const bool host_acc   = decltype(index_dataset)::accessor_type::is_host_type::value;
+    const bool device_acc = decltype(index_dataset)::accessor_type::is_device_type::value;
+    auto dataset          = raft::make_mdspan<const T, int64_t, row_major, host_acc, device_acc>(
+      index_dataset.data_handle(), extents);
```

Comment:
Yes, this is solely because of the matrix extent type. Unfortunately, because the IVF-Flat and IVF-PQ APIs were built against RAFT they also had to have a variable matrix extent type. I agree that we should in the end switch all matrix extent types to `int64_t`, but this should probably be done in an other PR. What do you think?

Reply:
Matrix extent type should be always int64_t. See my suggestions for the public API.